optuna/samplers/_grid.py

```diff
@@ -208,21 +208,27 @@ def _get_unvisited_grid_ids(self, study: Study) -> List[int]:
 
         # List up unvisited grids based on already finished ones.
         visited_grids = []
+        running_grids = []
 
         # We directly query the storage to get trials here instead of `study.get_trials`,
         # since some pruners such as `HyperbandPruner` use the study transformed
         # to filter trials. See https://github.com/optuna/optuna/issues/2327 for details.
         trials = study._storage.get_all_trials(study._study_id, deepcopy=False)
 
         for t in trials:
-            if (
-                t.state.is_finished()
-                and "grid_id" in t.system_attrs
-                and self._same_search_space(t.system_attrs["search_space"])
+            if "grid_id" in t.system_attrs and self._same_search_space(
+                t.system_attrs["search_space"]
             ):
-                visited_grids.append(t.system_attrs["grid_id"])
+                if t.state.is_finished():
+                    visited_grids.append(t.system_attrs["grid_id"])
+                elif t.state == TrialState.RUNNING:
+                    running_grids.append(t.system_attrs["grid_id"])
 
-        unvisited_grids = set(range(self._n_min_trials)) - set(visited_grids)
+        unvisited_grids = set(range(self._n_min_trials)) - set(visited_grids) - set(running_grids)
+
+        # If evaluations for all grids have been started, return grids that have not yet finished.
+        if len(unvisited_grids) == 0:
```

Comment:
[question]

It has to return grids of unfinished parameters because it will stop optimization leaving running trials if ` _get_unvisited_grid_ids` only returns unvisited trials. Is it right?

https://github.com/optuna/optuna/blob/10a9dfec5a819c8f9f441f7d704be15cc811a50e/optuna/samplers/_grid.py#L188-L189

If so, I think it would be useful to add a comment for developers. (IMO)
And I'm wondering it's straightforward to return `list(running_grids)` (sorry if I misunderstand).

Reply:
Thank you for your suggestion. I updated the comment.

`set(range(self._n_min_trials)) - set(visited_grids)` and `list(running_grids)` are not equivalent.
For example, in the following case, we should exclude grid 1 from the returned list.
- process 1 completed to evaluate grid 1.
- process 2 is running to evaluate grid 1.

Reply:
@not522 It makes sense, thank you for the explanation!